tasks/main.yml

```diff
@@ -20,34 +20,76 @@
     that:
       - mssql_accept_microsoft_odbc_driver_17_for_sql_server_eula | bool
       - mssql_accept_microsoft_cli_utilities_for_sql_server_eula | bool
-      - mssql_accept_microsoft_sql_server_2019_standard_eula | bool
+      - mssql_accept_microsoft_sql_server_standard_eula | bool
     fail_msg:
       - "You must accept EULA by setting the following variables to true:"
       - "mssql_accept_microsoft_odbc_driver_17_for_sql_server_eula"
       - "mssql_accept_microsoft_cli_utilities_for_sql_server_eula"
-      - "mssql_accept_microsoft_sql_server_2019_standard_eula"
+      - "mssql_accept_microsoft_sql_server_standard_eula"
+
+- name: Verify if the mssql_version variable is provided correctly
+  assert:
+    that: mssql_version == 2017 or mssql_version == 2019
+    fail_msg: >-
+      "The mssql_version variable must be set to either 2017 or 2019 and be of
+      type int."
+
+- name: Verify if the mssql_upgrade variable is provided correctly
+  fail:
+    msg:
+      - "You set mssql_upgrade to true and mssql_version to 2017."
+      - "If you want to upgrade, set the variables as such:"
+      - "mssql_version: 2019"
+      - "mssql_upgrade: true"
+  when:
+    - mssql_upgrade | bool
+    - mssql_version == 2017
+
+- name: Gather package facts
+  package_facts:
+    manager: auto
+
+- name: Verify if upgrade is needed
+  fail:
+    msg: >-
+      "You set mssql_upgrade to true, but your SQL Server is version 2019
+      already."
+  when:
+    - mssql_upgrade | bool
+    - ansible_facts.packages["mssql-server"][0]["version"] is search("^15.*")
 
 - name: Deploy the GPG key for Microsoft repositories
   rpm_key:
     key: https://packages.microsoft.com/keys/microsoft.asc
     state: present
 
-- name: Configure the Microsoft SQL Server 2019 repository
+- name: Remove the Microsoft SQL Server 2017 repository to upgrade to 2019
   yum_repository:
-    name: packages-microsoft-com-mssql-server-2019-gdr
-    description: Microsoft SQL Server 2019
+    name: packages-microsoft-com-mssql-server-2017-gdr
+    state: absent
+  when:
+    - mssql_version == 2019
+    - mssql_upgrade | bool
+
+- name: Configure the Microsoft SQL Server {{ mssql_version }} repository
+  yum_repository:
+    name: packages-microsoft-com-mssql-server-{{ mssql_version }}-gdr
+    description: Microsoft SQL Server {{ mssql_version }}
     baseurl: "{{ __mssql_server_repository }}"
     gpgcheck: true
+  when: >-
+    (__mssql_server_packages not in ansible_facts.packages) or
+    (mssql_upgrade | bool)
 
-- name: Ensure that the Microsoft SQL Server package is installed
+- name: Ensure the {{ __mssql_server_packages }} package
   package:
-    name: "{{ __mssql_server_packages }}"
-    state: present
+    name: "{{ __mssql_server_packages }}-1*"
```

Comment:
?

Reply:
Upgrade from mssql-server-14 to mssql-server-15 did not work for me without this workaround. When upgrading mssql-server with the `package` module to `state: latest`, Ansible threw dependency errors. I will look at it more closely later.

Reply:
When upgrading using just `name: "{{ __mssql_server_packages }}"`, the role fails with the following message:

=> {"changed": false, "failures": [], "msg": "Depsolve Error occured:
Problem:
- package python36-3.6.8-2.module+el8.1.0+3334+5cb623d7.x86_64 requires python3-pip, but none of the providers can be installed
- package python3-pip-9.0.3-18.el8.noarch requires platform-python-pip = 9.0.3-18.el8, but none of the providers can be installed
- package mssql-server-15.0.4138.2-1.x86_64 requires python3, but none of the providers can be installed
- cannot install both platform-python-pip-9.0.3-18.el8.noarch and platform-python-pip-9.0.3-19.el8.noarch
- cannot install the best update candidate for package mssql-server-14.0.3391.2-12.x86_64
- problem with installed package platform-python-pip-9.0.3-19.el8.noarch",
"rc": 1, "results": []}

I think this is because Ansible is confused by the mssql-server dependencies between versions mssql-server-14 and mssql-server-15. Ansible kinda wants to update mssql-server-14 to its latest version, and at the same gets carried away by upgrading to mssql-server-15, and pulling all the dependencies there.

This issue is reproducible with yum too. When mssql-server-14 is installed and the repository for mssql-server-15 is configured, running `yum update mssql-server` results in the same error. While running `yum install mssql-server` pulls the dependencies correctly.

When I use `name: "{{ __mssql_server_packages }}-1*"` in the module, Ansible upgrades mssql-server to mssql-server-15 successfully, which eliminates the error.

This workaround is confusing, I will change this to the following explicit solution instead:
```
package:
  name: >-
    {{ __mssql_server_packages + '-15*' if mssql_upgrade
    else __mssql_server_packages }}
```

Reply:
The usual workflow for doing a major upgrade is:
* ensure all packages are up-to-date from the old repos
* disable the old repos
* enable the new repos
* install/update the packages from the new repos

Reply:
Never mind, I just realized that the issue I was facing was caused by the older RHEL 8 guest image I run the tests against. I have now removed the weird workaround.

I have implemented updating all packages from the 2017 repo here:
https://github.com/linux-system-roles/mssql/pull/22/files#diff-3d0ff1709ca48add100327bb2a468e6c508fb92a159c64c4f99ad1df89d9bddeR77-R84

Is that what you expected?